Getting Started Guide: Linux (#17)
KhronosGroup/OpenCL-Guide: chapters/getting_started_linux.md

In this guide we'll be using latest (at the time of writing) Ubuntu 20.04 LTS. Installation for the most part will happen via [apt (Advanced Packaging Tool)](https://en.wikipedia.org/wiki/APT_(software)), the definitive command-line tool for installing software on Debian-base Linux distributives. It is installed with the system automatically.

_(NOTE: installation commands should be issued with root access, so that use of privilege-escalation by `sudo` is mandatory.)_

First update your system. Open your favorite shell in your favorite terminal and run:

``` bash
sudo apt update
sudo apt upgrade
```

### C/C++ compiler

Open your favorite shell in your favorite terminal and run:

``` bash
sudo apt install build-essential -y
```

This will install canonical GNU compiler set `g++/GNU 9.3`, debugger `gdb`, as well as some tools (e.g. `make`, `binutils`) and libraries for compilation and linking of programs.

Another popular choice is the usage of `Clang` compiler, based on intermediate representation of the C-like languages on `LLVM` (originally Low Level Virtual Machine). To use it open your favorite shell in your favorite terminal and run:

``` bash
sudo apt install clang -y
```

> This will install `Clang 10` and `LLVM 10`.

Reviewer: I'd consider removing this section (and any other mention of Clang) entirely.  While I'm a fan of Clang, I think this is unnecessary to create a "minimal and productive environment" and as far as I can tell no part of this guide uses Clang.

### Git

You most likely already have Git installed. If not, we'll install it, because this is what is needed to keep your repositories up-to-date. Open your favorite shell in your favorite terminal and run:

``` bash
sudo apt install git -y
```

### CMake

If you do not have CMake installed yet, we'll install it, as it's the primary supported build system, and it will make our builds much simpler. Open your favorite shell in your favorite terminal and run:

``` bash

> sudo apt install cmake ninja-build -y

Reviewer: Note, no part of this guide uses ninja.

Author: Removed.

```

### Visual Studio Code

> While IDEs are highly opinionated, for an IDE with widespared adoption and small footprint, VS Code has everything (and more) to get off the ground. It should be installed differently, though. The first way is to use GUI of `Ubuntu Software`, where you can search for `code` and install it. Whether you are a fan of command-line tools, use the second method: open your favorite shell in your favorite terminal and run

Reviewer 2: The English of this paragraph needs some help. Here are my suggestions:
```suggestion
While IDEs are highly opinionated, for an IDE with widespread adoption and small footprint, VS Code has everything (and more) to get off the ground. It can be installed in two ways. The first way is to use the `Ubuntu Software` GUI, where you can search for `code` and install it in one click. If you are a fan of command-line tools, use the second method: open your favorite shell in your favorite terminal and run
```

``` bash
sudo snap install --classic code
```

### OpenCL-SDK

To build native OpenCL applications, one will minimally need:
- C or C++ compiler
- The OpenCL headers
  - The C and optionally the C++ headers
- An Installable Client Driver (ICD) Loader
  - Shared objects library (libOpenCL.so)

>   - Export library (libOpenCL.a)

Reviewer: On Linux a separate export library is not needed and the same libOpenCL.so is used for both purposes.

Author: Yes, that was a mistake from our part. Fixed.

# Getting started with OpenCL on Ubuntu Linux

> OpenCL is included in all the main distributions of Linux. In this guide we choose Ubuntu as one of the most widespread variants, but all of the distributives have some sort of centralized system of package maintaining and OpenCL realizations can be installed from there.

Reviewer: ```suggestion
OpenCL is included in all the main distributions of Linux. In this guide we choose Ubuntu as one of the most widespread variants, but all of the distributions have some sort of centralized system of package maintaining and OpenCL realizations can be installed from there.
```

In this guide we're minimally going to use the following tools:

- The command-line

> - C/C++ compiler (gcc & Clang)

Reviewer: Minor: since we don't say anything about Clang anymore we may want to simplify this:
```suggestion
- C/C++ compiler
```

And depending on which ways one extends the bare minimum:

- CMake (Cross-platform Make)
- Git
- Vcpkg (Cross-platform pacakge management)
- Visual Studio Code

Steps will be provided to obtain a minimal and productive environment.

## Installation

> In this guide we'll be using latest (at the time of writing) Ubuntu 20.04 LTS. Installation for the most part will happen via [apt (Advanced Packaging Tool)](https://en.wikipedia.org/wiki/APT_(software)), the definitive command-line tool for installing software on Debian-base Linux distributives. It is installed with the system automatically.

Reviewer: ```suggestion
In this guide we'll be using latest (at the time of writing) Ubuntu 20.04 LTS. Installation for the most part will happen via [apt (Advanced Packaging Tool)](https://en.wikipedia.org/wiki/APT_(software)), the definitive command-line tool for installing software on Debian-base Linux distribution. It is installed with the system automatically.
```
